pkg/api/latest/codec.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package latest
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/conversion"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/runtime/serializer/json"
+	"k8s.io/kubernetes/pkg/runtime/serializer/recognizer"
+)
+
+var (
+	Codecs                CodecFactory
+	UniversalDeserializer runtime.Decoder
+
+	jsonSerializer = json.NewSerializer(conversion.DefaultMetaFactory, api.Scheme, runtime.ObjectTyperToTyper(api.Scheme))
+	yamlSerializer = json.NewYAMLSerializer(conversion.DefaultMetaFactory, api.Scheme, runtime.ObjectTyperToTyper(api.Scheme))
+)
+
+type serializerType struct {
+	AcceptContentTypes []string
+	ContentType        string
+	FileExtensions     []string
+	Serializer         runtime.Serializer
+}
+
+// allows other codecs to be enabled with compile time flags in their own files
+var serializers = []serializerType{
+	{
+		AcceptContentTypes: []string{"application/json"},
+		ContentType:        "application/json",
+		FileExtensions:     []string{"json"},
+		Serializer:         jsonSerializer,
+	},
+	{
+		AcceptContentTypes: []string{"application/yaml"},
+		ContentType:        "application/yaml",
+		FileExtensions:     []string{"yaml"},
+		Serializer:         yamlSerializer,
+	},
+}
+
+func init() {
+	Codecs = CodecFactory{
+		scheme:      api.Scheme,
+		serializers: serializers,
+	}
+	decoders := make([]runtime.Decoder, 0, len(serializers))
+	for _, d := range serializers {
+		decoders = append(decoders, d.Serializer)
+	}
+	UniversalDeserializer = recognizer.NewDecoder(decoders...)
+}
+
+type CodecFactory struct {
+	scheme      *runtime.Scheme
+	serializers []serializerType
+}
+
+type noopEncoder struct {
+	runtime.Decoder
+}
+
+func (e noopEncoder) EncodeToStream(obj runtime.Object, w io.Writer) error {
+	return fmt.Errorf("encoding is not allowed for this codec: %v", reflect.TypeOf(e.Decoder))
+}
+
+// UniversalDecoder returns a runtime.Decoder capable of decoding all known API objects in all known formats. Used
```

Comment:
Recognizer is a sniffer that peeks at the content stream looking for
sequences.  We can tell the difference between yaml, json, and protobuf
with magic number in the short term.  Not sure what to do for longer term
things.  Lots of our code depends on the "I expect to decode anything you
pass to me because I know about latest" which we need to break.


On Dec 1, 2015, at 7:08 PM, Daniel Smith <notifications@github.com> wrote:

In pkg/api/latest/codec.go
<https://github.com/kubernetes/kubernetes/pull/17922#discussion_r46360143>:

> +}
> +
> +type CodecFactory struct {
> +	scheme      *runtime.Scheme
> +	serializers []serializerType
> +}
> +
> +type noopEncoder struct {
> +	runtime.Decoder
> +}
> +
> +func (e noopEncoder) EncodeToStream(obj runtime.Object, w io.Writer) error {
> +	return fmt.Errorf("encoding is not allowed for this codec: %v", reflect.TypeOf(e.Decoder))
> +}
> +
> +// UniversalDecoder returns a runtime.Decoder capable of decoding all known API objects in all known formats. Used

How do you tell it of the format? Does it figure out the difference between
proto & json by itself?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17922/files#r46360143>.